File: app/backend/impl/ExternalCourseHandlerImpl.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package backend.impl;

import backend.models.Course;
import backend.models.Grade;
import backend.models.GradeScale;
import backend.models.Organisation;
import backend.models.User;
import backend.util.config.ConfigReader;
import com.fasterxml.jackson.databind.JsonNode;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import io.ebean.Ebean;
import java.net.MalformedURLException;
import java.net.URL;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;
import javax.inject.Inject;
import org.springframework.beans.BeanUtils;
import play.Logger;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;

public class ExternalCourseHandlerImpl implements ExternalCourseHandler {
    private static final String COURSE_CODE_PLACEHOLDER = "${course_code}";
    private static final String USER_ID_PLACEHOLDER = "${employee_number}";
    private static final String USER_LANG_PLACEHOLDER = "${employee_lang}";
    private static final boolean API_KEY_USED = ConfigFactory
        .load()
        .getBoolean("sitnet.integration.enrolmentPermissionCheck.apiKey.enabled");
    private static final String API_KEY_NAME = ConfigFactory
        .load()
        .getString("sitnet.integration.enrolmentPermissionCheck.apiKey.name");
    private static final String API_KEY_VALUE = ConfigFactory
        .load()
        .getString("sitnet.integration.enrolmentPermissionCheck.apiKey.value");
    private static final String USER_IDENTIFIER = ConfigFactory
        .load()
        .getString("sitnet.integration.enrolmentPermissionCheck.id");

    private static final DateFormat DF = new SimpleDateFormat("yyyyMMdd");

    private static final Logger.ALogger logger = Logger.of(ExternalCourseHandlerImpl.class);

    private static class RemoteException extends Exception {

        RemoteException(String message) {
            super(message);
        }
    }

    @FunctionalInterface
    private interface RemoteFunction<T, R> extends Function<T, R> {
        @Override
        default R apply(T t) {
            try {
                return exec(t);
            } catch (RemoteException | ParseException e) {
                throw new RuntimeException(e);
            }
        }

        R exec(T t) throws RemoteException, ParseException;
    }

    private final WSClient wsClient;
    private final ConfigReader configReader;

    @Inject
    public ExternalCourseHandlerImpl(WSClient wsClient, ConfigReader configReader) {
        this.wsClient = wsClient;
        this.configReader = configReader;
    }

    private Set<Course> getLocalCourses(String code) {
        return Ebean
            .find(Course.class)
            .where()
            .ilike("code", code + "%")
            .disjunction()
            .isNull("endDate")
            .gt("endDate", new Date())
            .endJunction()
            .disjunction()
            .isNull("startDate")
            .lt("startDate", new Date())
            .endJunction()
            .orderBy("code")
            .findSet();
    }

    @Override
    public CompletionStage<Set<Course>> getCoursesByCode(User user, String code) throws MalformedURLException {
        if (!configReader.isCourseSearchActive()) {
            return CompletableFuture.completedFuture(getLocalCourses(code));
        }
        // Hit the remote end for possible matches. Update local records with matching remote records.
        // Finally return all matches (local & remote)
        URL url = parseUrl(user.getOrganisation(), code);
        return downloadCourses(url)
            .thenApplyAsync(
                remotes -> {
                    remotes.forEach(this::saveOrUpdate);
                    Supplier<TreeSet<Course>> supplier = () -> new TreeSet<>(Comparator.comparing(Course::getCode));
                    return Stream
                        .concat(getLocalCourses(code).stream(), remotes.stream())
                        .collect(Collectors.toCollection(supplier));
                }
            );
    }

    @Override
    public CompletionStage<Collection<String>> getPermittedCourses(User user) throws MalformedURLException {
        URL url = parseUrl(user);
        WSRequest request = wsClient.url(url.toString().split("\\?")[0]);
        if (url.getQuery() != null) {
            request = request.setQueryString(url.getQuery());
        }
        if (API_KEY_USED) {
            request = request.addHeader(API_KEY_NAME, API_KEY_VALUE);
        }
        RemoteFunction<WSResponse, Collection<String>> onSuccess = response -> {
            JsonNode root = response.asJson();
            if (root.has("exception")) {
                throw new RemoteException(root.get("exception").asText());
            } else if (root.has("data")) {
                return StreamSupport
                    .stream(root.get("data").spliterator(), false)
                    .filter(c -> c.has("course_code"))
                    .map(c -> c.get("course_code").asText())
                    .collect(Collectors.toSet());
            } else {
                logger.warn("Unexpected content {}", root.asText());
                throw new RemoteException("sitnet_request_timed_out");
            }
        };
        return request.get().thenApplyAsync(onSuccess);
    }

    private void saveOrUpdate(Course external) {
        Ebean
            .find(Course.class)
            .where()
            .eq("code", external.getCode())
            .findOneOrEmpty()
            .ifPresentOrElse(
                local -> {
                    // Existing course
                    if (external.getCourseImplementation() != null) {
                        // update only those courses that specify an implementation
                        BeanUtils.copyProperties(external, local, "id", "objectVersion");
                        local.update();
                    }
                    external.setId(local.getId());
                },
                external::save
            );
    }

    private CompletionStage<List<Course>> downloadCourses(URL url) {
        WSRequest request = wsClient.url(url.toString().split("\\?")[0]);
        if (url.getQuery() != null) {
            request = request.setQueryString(url.getQuery());
        }
        RemoteFunction<WSResponse, List<Course>> onSuccess = response -> {
            int status = response.getStatus();
            if (status == Http.Status.OK) {
                return parseCourses(response.asJson());
            }
            logger.info("Non-OK response received for URL: {}. Status: {}", url, status);
            throw new RemoteException(String.format("sitnet_remote_failure %d %s", status, response.getStatusText()));
        };
        return request
            .get()
            .thenApplyAsync(onSuccess)
            .exceptionally(
                t -> {
                    logger.error("Connection error occurred", t);
                    return Collections.emptyList();
                }
            );
    }

    private static URL parseUrl(Organisation organisation, String courseCode) throws MalformedURLException {
        String urlConfigPrefix = "sitnet.integration.courseUnitInfo.url";
        Config config = ConfigFactory.load();
        String configPath = null;
        if (organisation != null && organisation.getCode() != null) {
            String path = String.format("%s.%s", urlConfigPrefix, organisation.getCode());
            if (config.hasPath(path)) {
                configPath = path;
            }
        }
        if (configPath == null) {
            String path = String.format("%s.%s", urlConfigPrefix, "default");
            if (config.hasPath(path)) {
                configPath = path;
            } else {
                throw new RuntimeException("sitnet.integration.courseUnitInfo.url holds no suitable URL for user");
            }
        }
        String url = ConfigFactory.load().getString(configPath);
        if (url == null || !url.contains(COURSE_CODE_PLACEHOLDER)) {
            throw new RuntimeException("sitnet.integration.courseUnitInfo.url is malformed");
        }
        url = url.replace(COURSE_CODE_PLACEHOLDER, courseCode);
        return new URL(url);
    }

    private static URL parseUrl(User user) throws MalformedURLException {
        if (USER_IDENTIFIER.equals("userIdentifier") && user.getUserIdentifier() == null) {
            throw new MalformedURLException("User has no identier number!");
        }
        String url = ConfigFactory.load().getString("sitnet.integration.enrolmentPermissionCheck.url");
        if (url == null || !url.contains(USER_ID_PLACEHOLDER)) {
            throw new MalformedURLException("sitnet.integration.enrolmentPermissionCheck.url is malformed");
        }
        String identifier = USER_IDENTIFIER.equals("userIdentifier") ? user.getUserIdentifier() : user.getEppn();
        url = url.replace(USER_ID_PLACEHOLDER, identifier).replace(USER_LANG_PLACEHOLDER, user.getLanguage().getCode());
        return new URL(url);
    }

    private List<GradeScale> getGradeScales(JsonNode src) {
        JsonNode node = src;
        List<GradeScale> scales = new ArrayList<>();
        if (node.has("gradeScale")) {
            node = node.get("gradeScale");
            for (JsonNode scale : node) {
                String type = scale.get("type").asText();
                Optional<GradeScale.Type> scaleType = GradeScale.Type.get(type);
                if (scaleType.isEmpty()) {
                    // not understood
                    logger.warn("Skipping over unknown grade scale type {}", type);
                } else if (scaleType.get().equals(GradeScale.Type.OTHER)) {
                    // This needs custom handling
                    if (!scale.has("code") || !scale.has("name")) {
                        logger.warn(
                            "Skipping over grade scale of type OTHER, required nodes are missing: {}",
                            scale.asText()
                        );
                        continue;
                    }
                    Long externalRef = scale.get("code").asLong();
                    GradeScale gs = Ebean.find(GradeScale.class).where().eq("externalRef", externalRef).findOne();
                    if (gs != null) {
                        scales.add(gs);
                        continue;
                    }
                    gs = new GradeScale();
                    gs.setDescription(GradeScale.Type.OTHER.toString());
                    gs.setExternalRef(externalRef);
                    gs.setDisplayName(scale.get("name").asText());
                    logger.info("saving scale " + externalRef);
                    gs.save();
                    for (JsonNode grade : scale.get("grades")) {
                        if (!grade.has("description")) {
                            logger.warn("Skipping over grade, required nodes are missing: {}", grade.asText());
                            continue;
                        }
                        Grade g = new Grade();
                        g.setName(grade.get("description").asText());
                        g.setGradeScale(gs);
                        // Dumb JSON API gives us boolean values as strings
                        boolean marksRejection =
                            grade.get("isFailed") != null &&
                            Boolean.parseBoolean(grade.get("isFailed").asText("false"));
                        g.setMarksRejection(marksRejection);
                        g.save();
                        gs.getGrades().add(g);
                    }
                    scales.add(gs);
                } else {
                    scales.add(Ebean.find(GradeScale.class, scaleType.get().getValue()));
                }
            }
        }
        return scales;
    }

    private Optional<Course> parseCourse(JsonNode node) throws ParseException {
        // check that this is a course node, response can also contain error messages and so on
        if (
            node.has("identifier") &&
            node.has("courseUnitCode") &&
            node.has("courseUnitTitle") &&
            node.has("institutionName")
        ) {
            Course course = new Course();
            if (node.has("endDate")) {
                Date endDate = DF.parse(node.get("endDate").asText());
                if (endDate.before(new Date())) {
                    return Optional.empty();
                }
                course.setEndDate(endDate);
            }
            if (node.has("startDate")) {
                Date startDate = DF.parse(node.get("startDate").asText());
                if (startDate.after(new Date())) {
                    return Optional.empty();
                }
                course.setStartDate(startDate);
            }
            course.setIdentifier(node.get("identifier").asText());
            course.setName(node.get("courseUnitTitle").asText());
            String code = node.get("courseUnitCode").asText();
            if (node.has("courseImplementation")) {
                String impl = node.get("courseImplementation").asText();
                course.setCourseImplementation(impl);
                course.setCode(String.format("%s_%s", code, impl));
            } else {
                course.setCode(code);
            }
            if (node.has("courseUnitLevel")) {
                course.setLevel(node.get("courseUnitLevel").asText());
            }
            if (node.has("courseUnitType")) {
                course.setCourseUnitType(node.get("courseUnitType").asText());
            }
            if (node.has("credits")) {
                course.setCredits(node.get("credits").asDouble());
            }
            String name = node.get("institutionName").asText();
            // TODO: how to identify (external) organisations. Maybe we need some "externalRef" for organisations as well?
            Organisation organisation = Ebean.find(Organisation.class).where().ieq("name", name).findOne();
            // TODO: should organisations preexist or not? As a safeguard, lets create these for now if not found.
            if (organisation == null) {
                organisation = new Organisation();
                organisation.setName(name);
                organisation.save();
            }
            course.setOrganisation(organisation);
            List<GradeScale> scales = getGradeScales(node);
            if (!scales.isEmpty()) {
                // For now support just a single scale per course
                course.setGradeScale(scales.get(0));
            }
            // in array form
            course.setCampus(getFirstChildNameValue(node, "campus").orElse(null));
            course.setDegreeProgramme(getFirstChildNameValue(node, "degreeProgramme").orElse(null));
            course.setDepartment(getFirstChildNameValue(node, "department").orElse(null));
            course.setLecturerResponsible(getFirstChildNameValue(node, "lecturerResponsible").orElse(null));
            course.setLecturer(getFirstChildNameValue(node, "lecturer").orElse(null));
            course.setCreditsLanguage(getFirstChildNameValue(node, "creditsLanguage").orElse(null));
            return Optional.of(course);
        }
        return Optional.empty();
    }

    private List<Course> parseCourses(JsonNode response) throws ParseException {
        List<Course> results = new ArrayList<>();
        if (response.get("status").asText().equals("OK") && response.has("CourseUnitInfo")) {
            JsonNode root = response.get("CourseUnitInfo");
            if (root.isArray()) {
                for (JsonNode node : root) {
                    parseCourse(node).ifPresent(results::add);
                }
            } else {
                parseCourse(root).ifPresent(results::add);
            }
        }
        return results;
    }

    private static Optional<String> getFirstChildNameValue(JsonNode json, String columnName) {
        if (json.has(columnName)) {
            JsonNode array = json.get(columnName);
            if (array.has(0)) {
                JsonNode child = array.get(0);
                if (child.has("name")) {
                    return Optional.of(child.get("name").asText());
                }
            }
        }
        return Optional.empty();
    }
}

File: app/services/exam/ExternalCourseHandlerImpl.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.exam

import database.EbeanQueryExtensions
import io.ebean.DB
import models.exam.{Course, Grade, GradeScale}
import models.facility.Organisation
import models.user.User
import org.apache.pekko.util.ByteString
import play.api.Logging
import play.api.libs.json.*
import play.api.libs.ws.{WSClient, WSResponse}
import play.mvc.Http
import schema.ExternalCourseValidator.{CourseUnitInfo, GradeScale as ExtGradeScale}
import security.BlockingIOExecutionContext
import services.config.ConfigReader

import java.net.*
import java.nio.charset.StandardCharsets
import java.time.format.DateTimeFormatter
import java.time.{Instant, ZoneOffset}
import javax.inject.Inject
import scala.collection.immutable.TreeSet
import scala.concurrent.Future
import scala.jdk.CollectionConverters.*
import scala.util.Try

class ExternalCourseHandlerImpl @Inject (
    private val wsClient: WSClient,
    private val configReader: ConfigReader,
    implicit val ec: BlockingIOExecutionContext
) extends ExternalCourseHandler
    with EbeanQueryExtensions
    with Logging:

  private val COURSE_CODE_PLACEHOLDER = "${course_code}"
  private val USER_ID_PLACEHOLDER     = "${employee_number}"
  private val USER_LANG_PLACEHOLDER   = "${employee_lang}"
  private val DF                      = DateTimeFormatter.ofPattern("yyyyMMdd")
  private val BOM = ByteString.fromArray(Array[Byte](0xef.toByte, 0xbb.toByte, 0xbf.toByte))

  private def getLocalCourses(code: String) = DB
    .find(classOf[Course])
    .where
    .ilike("code", code + "%")
    .disjunction
    .isNull("endDate")
    .gt("endDate", Instant.now())
    .endJunction
    .orderBy("code")
    .distinct
    .filter(c =>
      Option(c.startDate).isEmpty || configReader.getCourseValidityDate(
        c.startDate.toInstant
      ).isBefore(Instant.now())
    )

  override def getCoursesByCode(user: User, code: String): Future[Set[Course]] =
    if !configReader.isCourseSearchActive then Future(getLocalCourses(code))
    // Hit the remote end for possible matches. Update local records with matching remote records.
    // Finally, return all matches (local and remote)
    else
      val url = parseUrl(user.organisation, code)
      downloadCourses(url).map(externals =>
        externals.foreach(saveOrUpdate)
        TreeSet.empty[Course](
          using (a, b) => a.code.compareTo(b.code)
        ) ++ externals ++ getLocalCourses(code)
      )

  override def getPermittedCourses(user: User): Future[Set[String]] =
    val url = parseUrl(user)
    queryRequest(url)
      .get()
      .map(response =>
        parseResponseBody(response) match
          case Some(root) =>
            val nodes = root \ "data" \\ "course_code" ++ root \ "data" \\ "courseUnitCode"
            nodes.map(_.as[String]).toSet
          case None => Set.empty
      )
      .recover {
        case e: JsResultException =>
          logger.error(
            "Unable to parse permitted courses data: JSON structure did not match expected format",
            e
          )
          Set.empty
        case e: Exception =>
          logger.error(
            "Unable to check for permitted courses due to exception in network connection",
            e
          )
          Set.empty
      }

  private def saveOrUpdate(external: Course): Unit =
    DB.find(classOf[Course])
      .where
      .eq("code", external.code)
      .find match
      case Some(local) =>
        // Existing course
        if Option(external.courseImplementation).nonEmpty then
          // update only those courses that specify an implementation
          local.code = external.code
          local.name = external.name
          local.level = external.level
          local.credits = external.credits
          local.courseUnitType = external.courseUnitType
          local.identifier = external.identifier
          local.startDate = external.startDate
          local.endDate = external.endDate
          local.courseImplementation = external.courseImplementation
          local.creditsLanguage = external.creditsLanguage
          local.gradeScale = external.gradeScale
          local.lecturer = external.lecturer
          local.lecturerResponsible = external.lecturerResponsible
          local.institutionName = external.institutionName
          local.department = external.department
          local.degreeProgramme = external.degreeProgramme
          local.campus = external.campus
          local.courseMaterial = external.courseMaterial
          local.organisation = external.organisation
          local.update()
        external.id = local.id
      case _ => external.save()

  private def parseResponseBody(response: WSResponse): Option[JsValue] =
    val bytes = response.bodyAsBytes
    val bodyBytes =
      if bytes.length >= 3 && bytes.splitAt(3)._1 == BOM then
        logger.warn("BOM character detected in the beginning of response body")
        bytes.drop(3)
      else bytes
    val body = bodyBytes.utf8String.trim
    if body.startsWith("<") then
      logger.warn("Response is not JSON (e.g. HTML error page). Body starts with '<'.")
      None
    else
      Try(Json.parse(bodyBytes.toArray)).toOption match
        case None =>
          logger.warn("Response was not valid JSON (e.g. HTML error page).")
          None
        case some => some

  private def downloadCourses(url: URL) =
    queryRequest(url)
      .get()
      .map(response =>
        val status = response.status
        if status == Http.Status.OK then
          parseResponseBody(response) match
            case Some(root) => parseCourses(root).flatMap(parseCourse)
            case None       => Seq.empty
        else
          logger.info(s"Non-OK response received for URL: %url. Status: $status")
          Seq.empty
      )
      .recover {
        case e: JsResultException =>
          logger.error(
            "Unable to parse course data: JSON structure did not match expected format",
            e
          )
          Seq.empty
        case e: Exception =>
          logger.error("Unable to download course data due to exception in network connection", e)
          Seq.empty
      }

  private def parseCourses(root: JsValue): Seq[CourseUnitInfo] =
    (root \\ "CourseUnitInfo").flatMap { node =>
      node.validate[CourseUnitInfo] match
        case JsSuccess(cui, _) => Seq(cui)
        case JsError(_) =>
          node.validate[Seq[CourseUnitInfo]] match
            case JsSuccess(cuis, _) => cuis
            case JsError(errors) =>
              logger.warn(s"Failed to parse CourseUnitInfo: ${JsError.toJson(errors)}")
              Seq.empty
    }.toSeq

  private def parseCourse(cui: CourseUnitInfo): Option[Course] =
    (validateStart(cui.startDate), validateEnd(cui.endDate)) match
      case (Right(start), Right(end)) => // good to go
        val model = new Course
        model.startDate = start.map(java.util.Date.from).orNull
        model.endDate = end.map(java.util.Date.from).orNull
        model.identifier = cui.identifier
        model.name = cui.title
        model.code = cui.code
        model.courseImplementation = cui.implementation.orElse(cui.altImplementation).orNull
        model.level = cui.level.orNull
        model.courseUnitType = cui.`type`.orNull
        // This is interesting trying to pass an optional to Java's nullable number
        model.credits = if cui.credits.nonEmpty then cui.credits.get else null
        val org = DB.find(classOf[Organisation]).where.ieq("name", cui.institutionName).find match
          case None =>
            val org2 = new Organisation
            org2.name = cui.institutionName
            org2.save()
            org2
          case Some(org) => org

        model.organisation = org
        model.campus = cui.campus.flatMap(_.headOption).map(_.name).orNull
        model.degreeProgramme = cui.programme.flatMap(_.headOption).map(_.name).orNull
        model.department = cui.department.flatMap(_.headOption).map(_.name).orNull
        model.lecturerResponsible = cui.responsible.flatMap(_.headOption).map(_.name).orNull
        model.lecturer = cui.lecturer.flatMap(_.headOption).map(_.name).orNull
        model.creditsLanguage = cui.language.flatMap(_.headOption).map(_.name).orNull
        model.gradeScale = importScales(cui).headOption.orNull
        Some(model)
      case _ => None

  private def importScales(cui: CourseUnitInfo) =
    val (ext, loc) =
      cui.gradeScales.map(_.partition(_.`type` == "OTHER")).getOrElse((Seq.empty, Seq.empty))
    ext.flatMap(importScale) ++ loc.flatMap(gs =>
      DB.find(classOf[GradeScale]).where.eq("description", gs.`type`).find
    )

  private def importScale(gs: ExtGradeScale) = gs match
    case ExtGradeScale(Some(name), "OTHER", Some(code), Some(grades)) =>
      DB.find(classOf[GradeScale]).where.eq("externalRef", code).find.orElse {
        val model = new GradeScale
        model.description = GradeScale.Type.OTHER.toString
        model.externalRef = code
        model.displayName = name
        logger.info(s"saving scale $code")
        model.save()
        model.grades =
          grades.values
            .map(g =>
              val g2 = new Grade
              g2.name = g.description
              g2.marksRejection = g.isFailed
              g2.gradeScale = model
              g2.save()
              g2
            )
            .toSet
            .asJava

        Some(model)
      }
    case _ => None

  private def validateStart(date: Option[String]) = date match
    case None => Right(None)
    case Some(d) =>
      val instant = java.time.LocalDate.parse(d, DF).atStartOfDay(ZoneOffset.UTC).toInstant
      val limit   = configReader.getCourseValidityDate(instant)
      if limit.isAfter(Instant.now()) then Left("too soon") else Right(Some(instant))

  private def validateEnd(date: Option[String]) = date match
    case None => Right(None)
    case Some(d) =>
      val instant = java.time.LocalDate.parse(d, DF).atStartOfDay(ZoneOffset.UTC).toInstant
      if instant.isBefore(Instant.now()) then Left("too late") else Right(Some(instant))

  private def queryRequest(url: URL) =
    val host        = url.toString.split("\\?").head
    val queryString = Option(url.getQuery).getOrElse("")
    val params = queryString
      .split("&")
      .flatMap { part =>
        part.split("=", 2) match
          case Array(k, v) =>
            Some(URLDecoder.decode(k.trim, StandardCharsets.UTF_8) -> URLDecoder.decode(
              v,
              StandardCharsets.UTF_8
            ))
          case _ => None
      }
      .toSeq
    val request =
      if params.isEmpty then wsClient.url(host)
      else wsClient.url(host).withQueryStringParameters(params*)
    if configReader.isApiKeyUsed then
      val header = (configReader.getApiKeyName, configReader.getApiKeyValue)
      request.addHttpHeaders(header)
    else request

  private def parseUrl(organisation: Organisation, courseCode: String) =
    val urlConfigPrefix = "exam.integration.courseUnitInfo.url"
    val configPathForOrg =
      Option(organisation)
        .flatMap(o => Option(o.code))
        .flatMap(c => Option(s"$urlConfigPrefix.$c").filter(configReader.hasPath))
    val configPath = configPathForOrg.orElse {
      val path = String.format("%s.%s", urlConfigPrefix, "default")
      Some(path).filter(configReader.hasPath)
    }
    val path = configReader.getString(configPath.getOrElse(""))
    if (!path.contains(COURSE_CODE_PLACEHOLDER))
      throw new RuntimeException("exam.integration.courseUnitInfo.url is malformed")
    val url =
      path.replace(COURSE_CODE_PLACEHOLDER, URLEncoder.encode(courseCode, StandardCharsets.UTF_8))
    URI.create(url).toURL

  private def parseUrl(user: User) =
    if configReader.getPermissionCheckUserIdentifier == "userIdentifier" && Option(
        user.userIdentifier
      ).isEmpty
    then
      throw new MalformedURLException("User has no identifier number!")
    val url =
      Option(configReader.getPermissionCheckUrl).filter(_.contains(USER_ID_PLACEHOLDER))
    url match
      case None => throw new MalformedURLException(
          "exam.integration.enrolmentPermissionCheck.url is malformed"
        )
      case Some(url) =>
        val identifier = URLEncoder.encode(
          if (configReader.getPermissionCheckUserIdentifier == "userIdentifier")
            user.userIdentifier
          else user.eppn,
          StandardCharsets.UTF_8
        )
        val patched = url
          .replace(USER_ID_PLACEHOLDER, identifier)
          .replace(USER_LANG_PLACEHOLDER, user.language.code)
        URI.create(patched).toURL
